### PR TITLE
Dev Queue pull request to master

### PR DESCRIPTION
**Pull request by mcdonals**
Pull request to merge dev_queue into master. At this point Queue supports enque(), deque() and a max capacity. The QueueNode class supports the generic type signature for it's elements. 

JUnit classes created for both Queue class and QueueNode class and they provide 100% code coverage.

### CODE REVIEW

File: src/queue/Queue.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * of 10.
+ * @author Scott McDonald and Jon Cronin
+ */
+public class Queue {
+    private QueueNode head;
+    private QueueNode tail;
+    private int capacity;
+    private int size;
+    
+    // Define the return values the Queue class methods may use
```
Comment (jonjoncronin) [inline src/queue/Queue.java:21]: Ya - Enums? why was that a problem?

File: src/queue/Queue.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    public static final int Q_SUCCESS = 0;
+    public static final int Q_ERR_NULL = -1;
+    public static final int Q_ERR_FULL = -2;
+    
+    /**
+     * Default constructor. Sets head and tail to null, max capacity to 10 and
+     * current size to 0.
+     */
+    public Queue() {
```
Comment (jonjoncronin) [inline src/queue/Queue.java:31]: do we have a coding standard? do you really care if our code looks a little different?

File: src/queue/Queue.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    
+    /**
+     * Alternate constructor that allows caller to set the maximum capacity.
+     * If capacity is set to a negative value then we will simply let the queue
+     * grow until the system is out of heap.
+     * @param capacity 
+     */
+    public Queue(int capacity) {
+        this();
```
Comment (jonjoncronin) [inline src/queue/Queue.java:45]: I'm unfamiliar with this syntax? what does this(); line do?

File: src/queue/Queue.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        this.capacity = capacity;
+    }
+    
+    /**
+     * Returns a new node at the tail of the queue as long as their is space
```
Comment (jonjoncronin) [inline src/queue/Queue.java:50]: ...there is space...

File: src/queue/Queue.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        // the queue grow as big as needed, or until limited by the heap.
+        if ((capacity > 0) && (size == capacity)) {
+            return Q_ERR_FULL;
+        }
+        
+        if (head == null) {
+            head = node;
+            tail = node;
+        } else {
+            tail.setNextNode(node);
```
Comment (jonjoncronin) [inline src/queue/Queue.java:74]: Should you check that node.next is also null? You don't want someone creating a cycle on the queue.

File: src/queue/Queue.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    
+    /**
+     * Returns the capacity of the queue. Negative values represent unlimited
+     * capacity. In other words the queue is only limited by the size of the 
+     * heap.
+     * @return int capacity 
+     */
+    public int getCapacity() {
+        return capacity;
```
Comment (jonjoncronin) [inline src/queue/Queue.java:109]: I don't remember - do you need to make this line "return this.capacity" or is "this." redundant?

File: src/queue/QueueNode.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+ * @author Scott McDonald and Jon Cronin
+ */
+package queue;
+
+/**
+ * The QueueNode class is used to represent each element in a Queue.
+ * 
+ * @author Scott McDonald and Jon Cronin
+ */
+public class QueueNode<E> {
```
Comment (jonjoncronin) [inline src/queue/QueueNode.java:14]: What is this <> notation?

File: src/queue/QueueNode.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    
+    /**
+     * Sets the next node this node will point to
+     * @param next nextNode
+     */
+    public void setNextNode(QueueNode next) {
+        this.nextNode = next;
+    }
+}
```
Comment (jonjoncronin) [inline src/queue/QueueNode.java:51]: Where is the toString() and equals()?

File: src/queue/Queue.java [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    
+    /**
+     * Returns the current size of the queue; aka the number of elements in the
+     * queue.
+     * @return int size
+     */
+    public int getSize() {
+        return size;
+    }
+
```
Comment (jonjoncronin) [inline src/queue/Queue.java:120]: Where is the toString() and equals()?

File: test/queue/QueueTest.java [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+import org.junit.Test;
+import java.lang.Integer;
+
+/**
+ * Test class used to validate our simple Queue ADT.
+ * @author Scott McDonald and Jon Cronin
+ */
+public class QueueTest extends TestCase {
+    
+    // Default JUnit constructor created by Netbeans. Leaving for now.
```
Comment (jonjoncronin) [inline test/queue/QueueTest.java:20]: I tried using the automatic template that Netbeans creates as well for the stack implementation - I ended up modifying the code to mimic the test files we've already got.

File: test/queue/QueueTest.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        super(testName);
+    }
+
+    /**
+     * testQueue validates the default Queue constructor.
+     */
+    public void testQueue() {
+        Queue myQ = new Queue();
+        Queue myQSize = new Queue(20);
+        assertTrue(myQ.deque() == null);
```
Comment (jonjoncronin) [inline test/queue/QueueTest.java]: A pattern I've been following is to not include a function call in the assert(). This allows me to debug the test a little more easily as the result of the function is captured as a local variable - I can see what the value is a little better.

File: test/queue/QueueTest.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+        
+        // Verify that inserting a null object returns Q_ERR_NULL, leaving
+        // the Q unmodified
+        enQRetVal = myQ.enque(null);
+        assertTrue(enQRetVal == Queue.Q_ERR_NULL);
+        
+        // We have a default Q, so the capacity is 10, validate we can enqueue
+        // 10 nodes and that the 11th node will return Q_ERR_FULL
+        QueueNode<Integer> node;
+        for (int ii=0; ii<11; ii++) {
```
Comment (jonjoncronin) [inline test/queue/QueueTest.java:68]: Just a question about your boundary value of 11. Is there a reason you pick < 11 rather than <= 10?

**Comment by jonjoncronin** [comment]
lots of questions about syntax that I need you to teach me but nothing looks blatantly wrong.

File: src/queue/Queue.java (hunk 1 above)
Comment (mcdonals) [inline src/queue/Queue.java:21]: Basically my JUnit skills are still weak and enums in Java are actually mini classes. When i tried it as an enum type there were some functions that i needed to overload and test or jacoco wouldn't give me 100% test coverage anymore. So I abandoned the enum class type error return method for the time being. It's an enhancement I could look at adding again in the future.

File: src/queue/Queue.java (hunk 2 above)
Comment (mcdonals) [inline src/queue/Queue.java:31]: Yeah I was wondering if we should try and care. I'm fine either way. I've been trying to follow what's in "Core Java for the Impatient" which is on Safari. I think it's closer to the normal Java convention than the typical C convention, but I don't really mind either way.

File: src/queue/Queue.java (hunk 3 above)
Comment (mcdonals) [inline src/queue/Queue.java:45]: that calls the default Queue constructor from the overloaded constructor so you don't have to rewrite all the lines and only have to worry about the differences between the constructor calls. You can do the same thing with inherited classes and their constructors but then you use super() instead of this(). this is shorthand for this object instance, so from the object instance you are accessing the object instance.

File: src/queue/Queue.java (hunk 4 above)
Comment (mcdonals) [inline src/queue/Queue.java:50]: one would hope :)

File: src/queue/Queue.java (hunk 5 above)
Comment (mcdonals) [inline src/queue/Queue.java:74]: Not a bad idea. Added the check and a new error case: Q_ERR_NEXT_NODE. Added corresponding JUnit case too.

File: src/queue/Queue.java (hunk 6 above)
Comment (mcdonals) [inline src/queue/Queue.java:109]: It's redundant, but there are times where you have conflict between a local variable and an instance variable with the same name. For example the following wouldn't work:

public void setCapacity(int capacity) {
    capacity = capacity;
}

but this would
public void setCapacity(int capacity) {
    this.capacity = capacity;
}

File: src/queue/QueueNode.java (hunk 7 above)
Comment (mcdonals) [inline src/queue/QueueNode.java:14]: This is the java generic type notation. You can have as many generic types in a class as you need. Users can then pass in any object for that type. It is similar to what you did using "object" in linkedList, but different because you can do some type checking on the generic types at runtime. I think it is the preferred style for Java API classes. The collections classes like ArrayList all use it. Core Java for the Impatient covers them well in chapter 6 "Generic Programming".

File: src/queue/QueueNode.java (hunk 8 above)
Comment (mcdonals) [inline src/queue/QueueNode.java:51]: I'll add them in round two.

File: src/queue/Queue.java (hunk 9 above)
Comment (mcdonals) [inline src/queue/Queue.java:120]: Round two :)

File: test/queue/QueueTest.java (hunk 12 above)
Comment (mcdonals) [inline test/queue/QueueTest.java:68]: I just think it's easier to read, but that's just me :)

File: test/queue/QueueTest.java (hunk 11 above)
Comment (mcdonals) [inline test/queue/QueueTest.java]: Not a bad idea. I've updated all the asserts i had that made a function call directly in the assert.
